controllers/propagator/replication_utils.go

```diff
@@ -0,0 +1,184 @@
+package propagator
+
+import (
+	"context"
+
+	k8sdepwatches "github.com/stolostron/kubernetes-dependency-watches/client"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	appsv1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/placementrule/v1"
+
+	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	policiesv1beta1 "open-cluster-management.io/governance-policy-propagator/api/v1beta1"
+	"open-cluster-management.io/governance-policy-propagator/controllers/common"
+)
+
+// labelsForRootPolicy returns the labels for given policy
+func labelsForRootPolicy(plc *policiesv1.Policy) map[string]string {
+	return map[string]string{common.RootPolicyLabel: fullNameForPolicy(plc)}
+}
+
+// fullNameForPolicy returns the fully qualified name for given policy
+// full qualified name: ${namespace}.${name}
+func fullNameForPolicy(plc *policiesv1.Policy) string {
+	return plc.GetNamespace() + "." + plc.GetName()
+}
+
+// equivalentReplicatedPolicies compares replicated policies. Returns true if they match.
+func equivalentReplicatedPolicies(plc1 *policiesv1.Policy, plc2 *policiesv1.Policy) bool {
+	// Compare annotations
+	if !equality.Semantic.DeepEqual(plc1.GetAnnotations(), plc2.GetAnnotations()) {
+		return false
+	}
+
+	// Compare labels
+	if !equality.Semantic.DeepEqual(plc1.GetLabels(), plc2.GetLabels()) {
+		return false
+	}
+
+	// Compare the specs
+	return equality.Semantic.DeepEqual(plc1.Spec, plc2.Spec)
+}
+
+func (r *PolicyReconciler) buildReplicatedPolicy(
+	root *policiesv1.Policy, decision appsv1.PlacementDecision,
+) (*policiesv1.Policy, error) {
+	replicatedName := fullNameForPolicy(root)
+
+	replicated := root.DeepCopy()
+	replicated.SetName(replicatedName)
+	replicated.SetNamespace(decision.ClusterNamespace)
+	replicated.SetResourceVersion("")
+	replicated.SetFinalizers(nil)
+	replicated.SetOwnerReferences(nil)
+
+	labels := root.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	// Extra labels on replicated policies
+	labels[common.ClusterNameLabel] = decision.ClusterName
+	labels[common.ClusterNamespaceLabel] = decision.ClusterNamespace
+	labels[common.RootPolicyLabel] = replicatedName
+
+	replicated.SetLabels(labels)
+
+	var err error
+
+	replicated.Spec.Dependencies, err = r.canonicalizeDependencies(replicated.Spec.Dependencies, root.Namespace)
+	if err != nil {
+		return replicated, err
+	}
+
+	for i, template := range replicated.Spec.PolicyTemplates {
+		replicated.Spec.PolicyTemplates[i].ExtraDependencies, err = r.canonicalizeDependencies(
+			template.ExtraDependencies, root.Namespace)
+		if err != nil {
+			return replicated, err
+		}
+	}
+
+	return replicated, nil
+}
+
+func (r *PolicyReconciler) canonicalizeDependencies(
```

Comment:
I think a comment describing what this function does would be useful, since the formatted dependency array is not really used in this repo so it might be a bit confusing why we are changing it